### ISSUE TITLE
[T.NeuralAutoregressive] Abstract method to compute inverse transformation - NotImplementedError

### ISSUE THREAD

**Issue by syyunn** [pyro-ppl/pyro#3235]
### Issue Description

I am trying to use NeuralAutoregressive transform as following codes:

base_dist = dist.Normal(torch.zeros(1), torch.ones(1))

```
from pyro.nn import AutoRegressiveNN
import pyro.distributions.transforms as T

arn = AutoRegressiveNN(input_dim=1, hidden_dims=[50], param_dims=[16]*3)
naf_transform = T.NeuralAutoregressive(arn, hidden_units=16)
pyro.module("naf", naf_transform)

flow_dist = dist.TransformedDistribution(base_dist, [naf_transform])

# and then.. need to "train" based on "actual observations"
steps = 1000
X_obs = torch.tensor(X, dtype=torch.float) # X is output of modulo operation
print(X_obs.shape)
X_obs = torch.tensor(X, dtype=torch.float).unsqueeze(-1)
print(X_obs.shape)
optimizer = torch.optim.Adam(naf_transform.parameters(), lr=1e-2) # spline transform from T.Spline
for step in range(steps):
    optimizer.zero_grad()
    loss = -flow_dist.log_prob(X_obs).mean() # nll
    loss.backward()
    optimizer.step()
    flow_dist.clear_cache()

    if step % 200 == 0:
        print('step: {}, loss: {}'.format(step, loss.item()))
```

but I am getting error message as follows:

```
    179 def _inverse(self, y):
    180     """
    181     Abstract method to compute inverse transformation.
    182     """
--> 183     raise NotImplementedError
```

Does it mean that I can't use `NeuralAutoregressive` fully using pyro and need to implement inverse by myself?

**Comment by martinjankowiak** [comment]
@syyunn thanks for the question. in the future please ask these kinds of questions in our [forum](https://forum.pyro.ai/). github issues are intended for bug reports, feature requests, and the like.

as it says in the documentation of `NeuralAutoregressive`:

>   The inverse operation is not implemented. This would require numerical inversion, e.g., using a root finding method - a possibility for a future implementation.

even if this were implemented it would be quite slow. for this reason people usually use normalizing flows that are fast in the direction that is needed for training (e.g. if you're doing density estimation you need to be able to evaluate `log p(x)` efficiently for training data `x`.)

also note that there are other pytorch normalizing flows out there, e.g. [flowtorch](https://flowtorch.ai/). unfortunately we do not have the resources to build and maintain a fully feature complete normalizing flow library with lots of different flows
